### PR TITLE
feat(credentials): add support for vault KV-v2

### PR DESCRIPTION
**Pull request by louisruch**

### CODE REVIEW

**Comment by louisruch** [comment]
@mgaffney @tmessi this enables getting credentials from Vault KV 2 which has an embedded `data map[string]interface{}` in the secret Data response. Happy to discuss alternatives.

File: internal/credential/vault/internal/userpassword/userpassword.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		}
+	}
+
+	return
+}
+
+// kv2UserPass looks for the the usernameAttr and passwordAttr in the embedded
+// 'data' field within the data map.
+//
+// Additionaly it validates the data is in the expected KV-v2 format:
```
Comment (tmessi) [inline internal/credential/vault/internal/userpassword/userpassword.go:47]: I would phrase this so it is clear that the function uses this structure to determine that it is a kv2 secret and if this structure is not found, it returns.

File: internal/credential/vault/internal/userpassword/userpassword.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
+package userpassword
+
+type (
+	data         map[string]interface{}
+	userPassFunc func(sd data, usernameAttr, passwordAttr string) (string, string)
```
[the remark this answers is not in the document]
Comment (tmessi) [reply internal/credential/vault/internal/userpassword/userpassword.go:5]: Also, I would add a doc here describing this function and its expected behavior

File: internal/credential/vault/internal/userpassword/userpassword.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+			username = u
+		}
+	}
+	if p, ok := data[passwordAttr]; ok {
+		if p, ok := p.(string); ok {
+			password = p
+		}
+	}
+
+	return
```
Comment (tmessi) [inline internal/credential/vault/internal/userpassword/userpassword.go:89]: personally not a fan of naked returns, i think it is more readable and easier to verify if the returns are explicit

**Review by tmessi — APPROVED** [review]

**Pull request merged**